### PR TITLE
Controller V2 Libraries Interface

### PR DESCRIPTION
**Pull request by pwittrock**
[godoc link](http://35.188.196.116/pkg/github.com/kubernetes-sigs/kubebuilder/pkg/ctrl/)

### CODE REVIEW

File: pkg/ctrl/config.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import "k8s.io/client-go/rest"
+
+// Config provides a rest.Config that can be used to identify and authenticate with a Kubernetes cluster.
+type Config interface {
+	Config() rest.Config
+}
+
+var _ Config = ConfigFn(func() rest.Config { return rest.Config{} })
+
+// ConfigFn implements Config
```
Comment (apelisse) [inline pkg/ctrl/config.go]: Maybe use the usual `_ Config` trick?

File: pkg/ctrl/config.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+type ConfigFn func() rest.Config
```
Comment (apelisse) [inline pkg/ctrl/config.go]: `http` uses `HandlerFunc`[1], maybe `ConfigFunc`?

[1]: https://golang.org/pkg/net/http/#HandlerFunc

File: pkg/ctrl/controller.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ctrl
+
+// Reconcile enqueues ObjectKeys for a Reconcile to Reconcile.
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: That's 3 reconcile in one sentence, I better understand what this means ;-)

File: pkg/ctrl/controller.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+func (*Controller) WatchResource(Watchable, EventHandler) {
+	// TODO: Write this
+}
+
+// WatchChannel enqueues events from the channel
+func (*Controller) WatchChannel(chan interface{}, EventHandler) {
+	// TODO: Write this
+}
+
+// Start starts the controller watching for events and reconciling
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: That should say if this is blocking or not? Looks like your controller manager is blocking, but it's also assuming these are non-blocking, which is inconsistent? Make them all blocking and expect people to use goroutines, and use the channel to stop them?

File: pkg/ctrl/controller.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+type Controller struct {
+	// Reconcile Reconciles observed and desired object state within a cluster.
+	// Defaults to the DefaultReconcileFn
+	Reconcile Reconcile
+
+	// Name is the name that will appear in tracing, logging and monitoring metrics.
+	// Defaults to a generated value
+	Name string
+
+	// ControllerParallelism is the number of concurrent Reconciles
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: Pure nit, but you're not consistent in your end of sentence periods.

File: pkg/ctrl/controller.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	// Defaults to 1
+	ControllerParallelism int
+
+	// Stop is used to shutdown the Reconcile
+	// Defaults to a new channel
+	Stop <-chan struct{}
+
+	// BeforeReconcile is called before Reconciling an ObjectKey.
+	// Defaults to no-op
+	BeforeReconcile func(ObjectKey)
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: I'm not clear why this is needed. How is it different from doing something in my reconcile struct? (either by changing the struct itself, or by adding another struct that will intercept objects before they get sent to the actual reconcile function)

File: pkg/ctrl/controller.go [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+	// AfterReconcile is called after Reconciling an ObjectKey
+	// Defaults to no-op
+	AfterReconcile func(ObjectKey, err error)
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: Same as above. The simpler the struct the better I think

File: pkg/ctrl/controller.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+// WatchResource enqueues events from resource events.
+func (*Controller) WatchResource(Watchable, EventHandler) {
+	// TODO: Write this
+}
+
+// WatchChannel enqueues events from the channel
+func (*Controller) WatchChannel(chan interface{}, EventHandler) {
```
Comment (apelisse) [inline pkg/ctrl/controller.go]: I don't know the nuance between `WatchChannel` and `WatchResource`. Can channel be converted into resource or the other way around? Can we only provide one option and convert from one to the other?

File: pkg/ctrl/manager.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	controllers []*Controller
+}
+
+// RegisterController registers a controller to be started with Start.
+// The Stop channel will be set on any Controllers where it is unspecified.
+func (cm ControlleManager) RegisterController(c *Controller) {
+	cm.controllers = append(cm.controllers, c)
+}
+
+// Start starts all the controllers
```
Comment (apelisse) [inline pkg/ctrl/manager.go]: As discussed before, document that this is blocking.

File: pkg/ctrl/manager.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	// Default the Stop channel
+	if cm.Stop == nil {
+		cm.Stop = make(<-chan struct{})
+	}
+
+	// Start each controller
+	for _, c := range cm.controllers {
+		// Default the controller stop channel
+		if c.Stop == nil {
+			c.Stop = cm.Stop
```
Comment (apelisse) [inline pkg/ctrl/manager.go]: Each controller will fight to get the signal here, but only one will receive it. How do you know how many signal to send?

File: pkg/ctrl/map_resource.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+// ObjectToObjectKeys maps an object to a collection of keys to be enqueued
+type ObjectToObjectKeys interface {
+	Map(interface{}) []ObjectKey
+}
+
+var _ ObjectToObjectKeys = ObjectToObjectKeysFn(func(interface{}) []ObjectKey { return nil })
+
+// ObjectToObjectKeysFn implements ObjectToObjectKeys
+type ObjectToObjectKeysFn func(interface{}) []ObjectKey
```
Comment (apelisse) [inline pkg/ctrl/map_resource.go]: `Func`?

File: pkg/ctrl/map_resource.go [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+var _ EventHandler = MapResource{}
+
+// MapResource maps an object to a collection of keys
+type MapResource struct {
+	// ObjectToObjectKeys transforms the argument into a slice of keys to be reconciled
+	ObjectToKeys ObjectToObjectKeys
+
+	// Predicate implements Predicate, and is used to filter events before they are given to EventHandlers
+	// Defaults to accept all
+	Predicate
```
Comment (apelisse) [inline pkg/ctrl/map_resource.go]: Can the filter be implemented as a separate type of `MapResource` or `ObjectToObjectKeys`?

File: pkg/ctrl/reconcile.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+// Reconcile reconciles objects
+type Reconcile interface {
+	// Reconcile performs a full level-based reconciliation for the object referred to by the ObjectKey
+	Reconcile(ObjectKey) (ReconcileResult, error)
+}
+
+var _ Reconcile = ReconcileFn(func(ObjectKey) (ReconcileResult, error) { return ReconcileResult{}, nil })
+
+// ReconcileFn implements Reconcile
+type ReconcileFn func(ObjectKey) (ReconcileResult, error)
```
Comment (apelisse) [inline pkg/ctrl/reconcile.go]: `ReconcileFunc`?

File: pkg/ctrl/manager.go (hunk 10 above)
Comment (pwittrock) [reply pkg/ctrl/manager.go]: You close the channel and they will all get it.

File: pkg/ctrl/config.go (hunk 1 above)
Comment (pwittrock) [reply pkg/ctrl/config.go]: Are you suggesting something different from what is on line 26?

File: pkg/ctrl/config.go (hunk 2 above)
Comment (pwittrock) [reply pkg/ctrl/config.go]: Done

File: pkg/ctrl/controller.go (hunk 3 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Bad refactor :(

File: pkg/ctrl/controller.go (hunk 4 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Done.  Yes making them blocking.

File: pkg/ctrl/manager.go (hunk 9 above)
Comment (pwittrock) [reply pkg/ctrl/manager.go]: Done

File: pkg/ctrl/map_resource.go (hunk 11 above)
Comment (pwittrock) [reply pkg/ctrl/map_resource.go]: Done

File: pkg/ctrl/reconcile.go (hunk 13 above)
Comment (pwittrock) [reply pkg/ctrl/reconcile.go]: Done

File: pkg/ctrl/map_resource.go (hunk 12 above)
Comment (pwittrock) [reply pkg/ctrl/map_resource.go]: I don't understand.

Comment (apelisse) [reply pkg/ctrl/map_resource.go]: I'm wondering if the predicate has to be coupled (composed) to that layer, or if it can be inserted as a filter layer instead?

Comment (pwittrock) [reply pkg/ctrl/map_resource.go]: It could be separate.  Not sure what is more clean or consistent with other libraries.

File: pkg/ctrl/manager.go (hunk 10 above)
Comment (apelisse) [reply pkg/ctrl/manager.go]: I didn't even think about it ahah

File: pkg/ctrl/map_resource.go (hunk 12 above)
Comment (apelisse) [reply pkg/ctrl/map_resource.go]: Yeah, no strong opinion. I like to have each component do one thing, but the complexity might not be needed in that case.

File: pkg/ctrl/doc.go [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package ctrl provides libraries for implementing Controllers.
+
+Controller
```
Comment (apelisse) [inline pkg/ctrl/doc.go]: Random thought: shouldn't this be the documentation for the "Controller" struct, rather than in the top level doc part? Maybe true also for some of the sections you've created below. Though I also understand why you're doing it this way.

Comment (pwittrock) [reply pkg/ctrl/doc.go]: Hm.  I can put it there too.  There is a chicken/egg problem where users need to know what a Controller is before they will search it out.

File: pkg/ctrl/example_test.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+package ctrl_test
+
+import (
+	"fmt"
+
+	"github.com/kubernetes-sigs/kubebuilder/pkg/ctrl"
+	"k8s.io/api/apps/v1"
+)
+
+func ExampleController() {
```
Comment (apelisse) [inline pkg/ctrl/example_test.go]: You can also add docs to Examples (in case you forgot)

File: pkg/ctrl/doc.go (hunk 14 above)
Comment (apelisse) [reply pkg/ctrl/doc.go]: Maybe keep a brief introduction of how things are supposed to work together, so that people can then go below to get more information. Like...

> This package allows you to create and register your controllers so that you can reconcile blabla ...
> See example below.
>
> This example shows you how to register and start the reconcile loop:
> Example_Start() { ... }
>
> This example shows you how to set-up a more complicated Controller watcher ...
> Example_CombinesManyConcepts()

maybe?

File: pkg/ctrl/controller.go (hunk 8 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Maybe, but I don't think it would simplify things.  Would need to find some common interface for a channel and GVR.

File: pkg/ctrl/controller.go (hunk 6 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Removed.  It is useful for testing, but we can get at it another way.

**Comment by pwittrock** [comment]
Addressed a lot @apelisse comments.  A lot of renaming, and a bit of restructuring.

File: pkg/ctrl/example_test.go (hunk 15 above)
Comment (pwittrock) [reply pkg/ctrl/example_test.go]: Done.

File: pkg/ctrl/doc.go (hunk 14 above)
Comment (pwittrock) [reply pkg/ctrl/doc.go]: PTAL and let me know what you think.  I've changed the naming and docs quite a bit to address what are common miss conceptions about how controllers work.  The mechanics should be much more clear now.

File: pkg/ctrl/doc.go [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+2. Handle work items by Reconciling user requested state with the actual state in the cluster.
+
+The standard Controller workflow: Watch Event -> Enqueue ObjectKey(s) from Event -> Dequeue ObjectKey -> Reconcile
+
+Controllers often Watch multiple Resource types, such as Pods, ReplicaSets and Deployments, but they should
+enqueue only keys for a single Resource type.  e.g. For Deployment events, enqueue the key of the Deployment.
+For Pod and ReplicaSet Events, enqueue the key of the Deployment that owns the ReplicaSet (directly) or Pod
+(transitively through a ReplicaSet).
+
+Import Concept:
```
Comment (apelisse) [inline pkg/ctrl/doc.go]: did you mean "important"?

File: pkg/ctrl/doc.go [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+		ctrl.Enqueue{},
+	)
+
+Resource
+
+Resource implements Watchable and contains the Group, Version, and Resource of an API to Watch for Events.
+
+	ctrl.Resource{Group: "core", Version: "v1", Resource: "pods"}
+
+Channel
```
Comment (apelisse) [inline pkg/ctrl/doc.go]: You haven't really introduced channel before that, and I honestly don't understand how they would be useful by reading the description below?

File: pkg/ctrl/doc.go [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+	  controller := &ctrl.Controller{Name: "pod-controller", Reconcile: ctrl.DefaultReconcileFunc}
+	  ctrl.Register(controller)
+
+	  controller.Watch(
+	    ctrl.Resource{Group: "core", Version: "v1", Enqueue: "pods"},
+	    ctrl.Enqueue{},
+	  )
+	}
+
+Controllers
```
Comment (apelisse) [inline pkg/ctrl/doc.go]: this section is very useful, and definitely helps me understand what's going on. It is still pretty complicated in nature (e.g. `Watch Event -> Enqueue ObjectKey(s) from Event -> Dequeue ObjectKey -> Reconcile` is not obvious ...), and thus hard to explain in plain english.

I think this overall section would be extremely useful if you could explain using a good example. As in, "OK let's say we want to build a controller that does X (scale number of pods maybe?), here are the concepts we need to know and why, and how we use them.".

The examples that you give are all mostly empty and thus hard to relate to or understand ("right, I know how to do an empty version of this, but why do I even need it since it's empty?")

Then potentially, you can use some of the other examples to show some nuance "Let's say we want to do it a different way, we could use this feature".

Comment (apelisse) [reply pkg/ctrl/doc.go]: I don't think I really understand what an `Event` is.

Comment (apelisse) [reply pkg/ctrl/doc.go]: I don't see any definition of what an `ObjectKey` is either

File: pkg/ctrl/controller.go [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"github.com/kubernetes-sigs/kubebuilder/pkg/ctrl/eventhandler"
+	"github.com/kubernetes-sigs/kubebuilder/pkg/ctrl/predicate"
+	"github.com/kubernetes-sigs/kubebuilder/pkg/ctrl/reconcile"
+	"github.com/kubernetes-sigs/kubebuilder/pkg/ctrl/source"
+)
+
+// Controllers are work queues that watch for changes to objects (i.e. Create / Update / Delete Events) and
+// then make changes to the cluster in response (i.e. Reconcile the object).
```
Comment (perotinus) [inline pkg/ctrl/controller.go]: nit: maybe replace "cluster" with something more generic? I expect that people will use this library to write controllers against control planes that are not for kubernetes clusters.

File: pkg/ctrl/controller.go [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+type Controller struct {
+	// Name is the name of the Controller that will appear in tracing, logging and monitoring metrics.
+	// Name is required.
+	Name string
+
+	// Reconcile Reconciles observed and desired object state within a cluster.  Defaults to the DefaultReconcileFunc.
+	Reconcile reconcile.Reconcile
+
+	// ControllerParallelism is the maximum number of concurrent Reconciles which will be run. Defaults to 1.
+	ControllerParallelism int
```
Comment (perotinus) [inline pkg/ctrl/controller.go]: nit: call this MaxConcurrentReconciles?

File: pkg/ctrl/controller.go [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+)
+
+// Controllers are work queues that watch for changes to objects (i.e. Create / Update / Delete Events) and
+// then make changes to the cluster in response (i.e. Reconcile the object).
+type Controller struct {
+	// Name is the name of the Controller that will appear in tracing, logging and monitoring metrics.
+	// Name is required.
+	Name string
+
+	// Reconcile Reconciles observed and desired object state within a cluster.  Defaults to the DefaultReconcileFunc.
```
Comment (perotinus) [inline pkg/ctrl/controller.go]: same here: replace "within a cluster" with something else more generic, or remove it.

File: pkg/ctrl/controller_manager.go [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+limitations under the License.
+*/
+
+package ctrl
+
+// DefaultControllerManager is the default ControllerManager.
+var DefaultControllerManager = &ControlleManager{}
+
+// ControllerManager initializes and starts Controllers.
+type ControlleManager struct {
```
Comment (perotinus) [inline pkg/ctrl/controller_manager.go]: Could you put some documentation somewhere that explains how you expect the controller manager to be used? Are you expecting people to generally be writing multiple controllers that are bundled in the controller manager? It seems to me like a more-common use case will be to have a CRD with one controller associated with it, in which case having the controller manager seems like it adds some extra machinery that's not necessary.

Comment (perotinus) [inline pkg/ctrl/controller_manager.go]: Nit: ControllerManager

File: pkg/ctrl/doc.go [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+	1. Watch one or more Sources of Events (i.e. changes to Kubernetes objects resulting from Create / Update / Delete operations).
+
+	  1.1 EventHandlers enqueue ReconcileRequests (containing the Names and Namespaces of objects to Reconcile).
+
+	2. Dequeue a ReconcileRequest and Reconcile an object.
+
+	  2.1 Read the object referred to in the request as well as any related objects (e.g. owned objects).
+
+	  2.2 Create or update Kubernetes objects as needed to execute the desired state specified in the object.
```
Comment (perotinus) [inline pkg/ctrl/doc.go]: ... Kubernetes objects and/or external state...

File: pkg/ctrl/controller_manager.go (hunk 22 above)
Comment (pwittrock) [reply pkg/ctrl/controller_manager.go]: Done

File: pkg/ctrl/doc.go (hunk 23 above)
Comment (pwittrock) [reply pkg/ctrl/doc.go]: Done

File: pkg/ctrl/controller.go (hunk 20 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Done

**Comment by pwittrock** [comment]
PTAL.

File: pkg/ctrl/controller.go (hunk 19 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Done

File: pkg/ctrl/controller.go (hunk 21 above)
Comment (pwittrock) [reply pkg/ctrl/controller.go]: Done

**Comment by pwittrock** [comment]
Moving this to v2 branch
